### ISSUE TITLE
memory problems with large attachment download

### ISSUE THREAD

**Issue by ghost** [makumba/makumba#173]
**Reported by @cristianbogdan on 28 Mar 2003 14:37 UTC**
Farry says:

An OutofMemory error that occurded today (28/03/2003) might be related to this
bug. I was downloading a 8,5MB file from the archives, which hanged halfway.
Soon after, PA was no longer accesible, and /tomcat/logs/server-output.txt
revealed OuM error.
The problem Cristi describes in 
bug 19 comment 4 might be the
cause of this failure.
A solution - if working with streams is not an option - might be to first check
if there is enough memory available to allocate for the db operation before
actually allowing the operation.

Stefan says:

I was with Stjepan today in a printshop (10:30), _trying_ to download the
same 8.5meg file (to print the color preview), but the progress bar didn't
even start to move, it was just hanging there at 0% instead. However there
was no problem downloading one version prior to that one (2megs or so).

would be really good to have some other indication that memory is low than
just being there, waiting for a miracle :).

Migrated-From: http://trac.makumba.org/ticket/373

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 28 Mar 2003 14:49 UTC**
these problems manifest themselves both at upload (bug 103) and download. as
indicated in bug 103, at upload we can reduce the memory.

but about download, the problem is more serious because the allocation takes
place in mysql/J (in this version and in all previous versions, so the new
version asked for in bug 19 is not to blame). of course, mysql/J allocates at
upload as well but there we (makumba) allocate too, so fixing bug 103 may half
the upload allocation

now we cannot fix mysql/J unless we get deeply involved in that project which we
should not. but checking the available memory and refusing the download if it's
too low may be an idea

what should happen is that
- the large attachment (blob) is read from the db (=> allocation inside mysql/J)
- the attachment is transformed into an org.makumba.Text and _streamed_ (i'm not
sure if via the disk or not) to the response.

if we could somehow stream from the db to the disk (or directly to the
response), that would be nice. the new mysql/J may provide support for that
which didn't exist in its old versions. it may be e.g. possible to just read a
part of the attachment, resulting at least in a smaller allocation

in short, i think somebody should experiment with this, print memory footprints
during upload and download, etc

the classes involved (which in turn involve other classes) are org.makumba.Text
and org.makumba.db.sql.textManager. also the download code (written in JSP)
should be looked at; the JSP is a nice place where one could print memory
footprints to the standard output.

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 28 Mar 2003 14:54 UTC**
speaking of memory checks, if the check fails (i.e. r* attachment_size >
free_mem, where r is a constant determined experimentally as indicated in
comment 2) it may be possible to order a System.gc() and wait a bit

the page could return a meta-refresh with a text "not enough memory for the
download, you will be redirected in 2 minutes, please wait". maybe that can also
be done for uploads, but downloads are more critical since they happen more often

**Comment by ghost** [comment]
**Modified by @manuelbernhardt on 28 Jun 2010 07:22 UTC**

**Comment by ghost** [comment]
**Modified by @manuelbernhardt on 28 Jun 2010 07:36 UTC**

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 19 Mar 2013 02:56 UTC**
a best-effort was made by allowing attachments to be stored in files, etc
